activation/activation_test.go

```diff
@@ -271,48 +159,87 @@ func assertLastAtx(r *require.Assertions, posAtx, prevAtx *types.VerifiedActivat
+) (*types.ActivationTx, error) {
 	t.Helper()
-	net.lastTransmission = nil
-
-	nipostBuilderMock.mu.Lock()
-	nipostBuilderMock.buildNIPostFunc = func(challenge *types.PoetChallenge) (*types.NIPost, time.Duration, error) {
-		builtNIPost = true
-		layerClockMock.mu.Lock()
-		layerClockMock.currentLayer = layerClockMock.currentLayer.Add(buildNIPostLayerDuration)
-		layerClockMock.mu.Unlock()
-		hash, err := challenge.Hash()
-		require.NoError(t, err)
-		return newNIPostWithChallenge(hash, poetBytes), 0, nil
-	}
-	nipostBuilderMock.mu.Unlock()
-
-	layerClockMock.mu.Lock()
-	layerClockMock.currentLayer = clockEpoch.FirstLayer().Add(3)
-	layerClockMock.mu.Unlock()
-
-	err = b.PublishActivationTx(context.Background())
 
-	nipostBuilderMock.mu.Lock()
-	nipostBuilderMock.buildNIPostFunc = nil
-	nipostBuilderMock.mu.Unlock()
-
-	return net.lastTransmission != nil, builtNIPost, err
+	publishEpoch := posAtxLayer.GetEpoch() + 1
+	tab.mclock.EXPECT().GetCurrentLayer().DoAndReturn(
+		func() types.LayerID {
+			return *currLayer
+		}).AnyTimes()
+	tab.mhdlr.EXPECT().GetPosAtxID().Return(posAtxId, nil)
+	tab.mclock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(
+		func(got types.LayerID) time.Time {
+			require.Equal(t, publishEpoch.FirstLayer(), got)
+			return time.Now().Add(100 * time.Second)
+		})
+	lastOpts := DefaultPostSetupOpts()
+	tab.mpost.EXPECT().LastOpts().Return(&lastOpts).AnyTimes()
+	tab.mnipost.EXPECT().BuildNIPost(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(
+		func(_ context.Context, challenge *types.PoetChallenge, _ time.Time) (*types.NIPost, time.Duration, error) {
+			*currLayer = currLayer.Add(buildNIPostLayerDuration)
+			return newNIPostWithChallenge(challenge.Hash(), []byte("66666")), 0, nil
+		})
+	ch := make(chan struct{}, 1)
+	close(ch)
+	tab.mclock.EXPECT().AwaitLayer(publishEpoch.FirstLayer()).DoAndReturn(
+		func(got types.LayerID) chan struct{} {
+			// advance to publish layer
+			if currLayer.Before(got) {
+				*currLayer = got
+			}
+			return ch
+		})
+	var built *types.ActivationTx
+	tab.mpub.EXPECT().Publish(gomock.Any(), pubsub.AtxProtocol, gomock.Any()).DoAndReturn(
+		func(_ context.Context, _ string, got []byte) error {
+			gotAtx, err := types.BytesToAtx(got)
+			require.NoError(t, err)
+			built = gotAtx
+			require.NoError(t, built.CalcAndSetID())
+			if publishErr != nil {
+				return publishErr
+			}
+			vatx, err := built.Verify(0, 1)
+			require.NoError(t, err)
+			require.NoError(t, atxs.Add(tab.cdb, vatx, time.Now()))
+			return nil
+		})
+	never := make(chan struct{}, 1)
+	if publishErr == nil {
+		tab.mhdlr.EXPECT().AwaitAtx(gomock.Any()).Return(ch)
+		tab.mclock.EXPECT().AwaitLayer((publishEpoch + 2).FirstLayer()).Return(never)
+		tab.mhdlr.EXPECT().UnsubscribeAtx(gomock.Any()).DoAndReturn(
+			func(got types.ATXID) {
+				require.Equal(t, built.ID(), got)
+			})
+	} else {
+		tab.mhdlr.EXPECT().AwaitAtx(gomock.Any()).Return(never)
+		tab.mhdlr.EXPECT().UnsubscribeAtx(gomock.Any())
+	}
```

Comment:
I'd rather have the test-specific mock expectations in the test body rather than a generic, one-fit-all helper. It might quickly become a huge pain for somebody willing to write a test that doesn't fit these expectations exactly.

Reply:
agreed. thanks for the suggestion. done